Fix some compile warnings and errors on an up-to-date system

gatecat opened the pull request:
This enables OpenROAD to build with a recent compiler; without hitting warnings that become errors due to -Werror (the first commit, the `free` on a class member, is a particularly concerning issue...)

Not sure if all of these will be of interest as-is but hopefully at least some of this is useful!

maliberty commented:
What compiler reports these problems?

gatecat commented:
GCC 11.1.0

rovinski commented:
I still believe `-Werror` should only be enabled for CI builds for this reason.

gatecat commented:
> I still believe `-Werror` should only be enabled for CI builds for this reason.

Yeah, that's how I have nextpnr set up too because of similar concerns.

Note that 4edea48 is fixing a genuine error for me, not a warning-turned-error, so would still be needed and some of the other free-related ones might be worth looking at as it seems like the compiler has found genuine concerns.

maliberty commented:
Note that there is ALLOW_WARNINGS cmake option if needed.

maliberty commented:
What version of spdlog are you using?  I don't see any such assertion in the current version.

maliberty commented on src/OpenRoad.cc:
```diff
@@ -179,7 +179,6 @@ OpenRoad::~OpenRoad()
 void
 deleteAllMemory()
 {
-  delete OpenRoad::openRoad();
```
Why remove this?

gatecat replied:
It returns a pointer to a static object; which is not something that should be deleted afaik

maliberty replied:
Agreed

maliberty commented on src/drt/src/db/tech/frConstraint.h:
```diff
@@ -426,7 +426,7 @@ class frMinStepConstraint : public frConstraint
         "insideCorner {} outsideCorner {} step {} maxEdges {}",
         minStepLength,
-        minstepType,
+        int(minstepType),
```
Is this an spdlog version issue?  It generates no error in our build.

gatecat replied:
I'm on spdlog 1.9.2

from past experience; different compiler versions have different policies on whether or not enums can be implicitly converted to ints, that might be the problem here rather than something in spdlog

maliberty commented on src/grt/src/fastroute/CMakeLists.txt:
```diff
@@ -60,3 +60,9 @@ target_link_libraries(FastRoute4.1
     Boost::boost
 )
+
+# False positive? inside Boost
+# error: ‘void* __builtin_memset(void*, int, long unsigned int)’ offset 0 is out of the bounds [0, 0] [-Werror=array-bounds]
+target_compile_options(FastRoute4.1
+  PRIVATE $<$<CXX_COMPILER_ID:GNU>:-Wno-array-bounds>
+)
```
What version of boost?

gatecat replied:
1.78.0-1

rovinski commented:
> Note that [4edea48](https://github.com/The-OpenROAD-Project/OpenROAD/commit/4edea486af07f8ea7f9c0d6d00bf5eeb9d1c3761) is fixing a genuine error for me, not a warning-turned-error, so would still be needed and some of the other free-related ones might be worth looking at as it seems like the compiler has found genuine concerns.

Realistically it would be more helpful if these enums had a `toString()` function so that they can be printed.

gatecat commented on src/gui/src/layoutViewer.cpp:
```diff
@@ -196,10 +196,10 @@ class GuiPainter : public Painter
                                     QPoint(rect.xMax(), rect.yMax())),
-                              roundX,
-                              roundY);
+                              (rect.xMax() - rect.xMin()) * (roundX / 200.0f),
+                              (rect.yMax() - rect.yMin()) * (roundY / 200.0f));
```
The old API takes a percentage; the new one takes a radius so I was using (length)/2 as the radius for 100% and then scaling proportionality for anything below 100%, if that makes sense.

maliberty replied:
It does and a comment to that effect would help.

maliberty commented:
I fixed the issue addressed in 4edea486af07f8ea7f9c0d6d00bf5eeb9d1c3761 in https://github.com/The-OpenROAD-Project/OpenROAD/pull/1542/files in a fashion I prefer (operator<< not int cast).  Please revert that commit from this PR and re-merge master to this PR.

gatecat commented:
Sorry, this unfortunately ended up forgotten all too long as life got in the way in various forms - I've finally rebased it and addressed some comments although a new issue appeared with a recent `fmt` version - I've fixed it in [afb3313](https://github.com/The-OpenROAD-Project/OpenROAD/pull/1534/commits/afb3313bc6a3190aeae2c051d4d78c6af16c94c7) but I can see there are other approaches that could be taken there.

gatecat commented:
Seems like my fix in [afb3313](https://github.com/The-OpenROAD-Project/OpenROAD/pull/1534/commits/afb3313bc6a3190aeae2c051d4d78c6af16c94c7) was no good because it breaks older `fmt` versions... not sure what you think is the best approach to supporting everything here.

maliberty commented:
The fmt change has been discussed in various PRs.  I think FMT_DEPRECATED_OSTREAM is the best solution until fmt fixes https://github.com/fmtlib/fmt/issues/3088
